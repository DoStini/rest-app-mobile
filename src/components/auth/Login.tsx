import {
  TextInput,
  View,
  StyleSheet,
  Text,
  Keyboard,
  TouchableWithoutFeedback,
} from "react-native";
import FormStyles from "../../styles/Forms";
import Button from "../Button";
import TextStyles from "../../styles/Text";
import { Formik } from "formik";
import * as Yup from "yup";
import useAuth from "../../hooks/useAuth";

const LoginSchema = Yup.object().shape({
  email: Yup.string().email("Invalid email").required("Email is required"),
  password: Yup.string().required("Password is required"),
});

const Login = () => {
  const { fetchingLogin, errorLogin, login } = useAuth();

  return (
    <TouchableWithoutFeedback onPress={() => Keyboard.dismiss()}>
      <View style={styles.container}>
        <Text style={TextStyles.h1}>Welcome to Resty</Text>
        <Text style={{ ...TextStyles.body, paddingBottom: 30, paddingTop: 15 }}>
          Please authenticate to access the platform
        </Text>

        <View style={styles.formComponent}>
          <Formik
            initialValues={{ email: "", password: "" }}
            validationSchema={LoginSchema}
            onSubmit={(values) => login(values.email, values.password)}
          >
            {({ handleChange, handleBlur, handleSubmit, errors, values }) => (
              <>
                <TextInput
                  style={FormStyles.input}
                  placeholder="Email"
                  value={values.email}
                  testID="email"
                  onChangeText={handleChange("email")}
                  onBlur={handleBlur("email")}
                />
                {errors.email && (
                  <Text style={FormStyles.error}>{errors.email}</Text>
                )}

                <TextInput
                  style={FormStyles.input}
                  value={values.password}
                  testID="password"
                  placeholder="Password"
                  secureTextEntry
                  onChangeText={handleChange("password")}
                  onBlur={handleBlur("password")}
                />
                {errors.password && (
                  <Text style={{ ...FormStyles.error, paddingBottom: 12 }}>
                    {errors.password}
                  </Text>
                )}

                {errorLogin && (
                  <Text style={{ ...FormStyles.error, paddingBottom: 12 }}>
                    Invalid email or password
                  </Text>
                )}

                <Button
                  testID="submit"
                  text="Login"
                  onPress={handleSubmit}
                  loading={fetchingLogin}
                />
              </>
            )}
          </Formik>
        </View>
      </View>
    </TouchableWithoutFeedback>
  );
};

export default Login;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    maxWidth: 400,
  },
  formComponent: {
    width: "80%",
    maxWidth: 400,
  },
});
