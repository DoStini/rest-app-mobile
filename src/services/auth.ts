import { setItemAsync } from "expo-secure-store";
import api from "./axios";

export const login = async (email: string, password: string) => {
  const response = await api.post("/proxy/auth", {
    username: email,
    password,
  });

  const token = response.data.accessToken;

  await setItemAsync("token", token);

  return response.data;
};
