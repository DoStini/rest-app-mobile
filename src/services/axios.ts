import axios from "axios";
import { getItemAsync } from "expo-secure-store";

const api = axios.create({
  baseURL: process.env.EXPO_PUBLIC_API_BASE_ROUTE,
});

api.interceptors.request.use(
  async (config) => {
    const token = await getItemAsync("token");
    if (token) {
      config.headers["Authorization"] = `Bearer ${token}`;
    }
    return config;
  },
  (error) => Promise.reject(error)
);

export const getFetcher = async (url: string) => {
  try {
    const response = await api.get(url);
    return response.data;
  } catch (error) {
    console.error("Error fetching data:", error);
  }
};

export default api;
