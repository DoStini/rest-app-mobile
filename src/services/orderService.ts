import api from "./axios";

export const fetchHistory = async () => {
  try {
    const response = await api.get("/orders/closed");
    return response.data.orders;
  } catch (error) {
    console.error("Error fetching categories:", error);
    throw error;
  }
};

export const fetchOrderById = async (id: string) => {
  try {
    const response = await api.get(`/orders/${id}`);
    return response.data;
  } catch (error) {
    console.error("Error fetching categories:", error);
    throw error;
  }
};

/*
TODO add right order id. Or dunno is there an api call to fetch all catefories without order id?
export const fetchCategories = async () => {
  try {
    const response = await api.get("/orders/471/categories");
    return response.data.categories;
  } catch (error) {
    console.error("Error fetching categories:", error);
    throw error;
  }
};*/
