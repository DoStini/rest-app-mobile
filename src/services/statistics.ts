import api from "./axios";

export const mainStatistics = async () => {
  try {
    const response = await api.get("/statistics");
    return response.data;
  } catch (error) {
    console.error("Error fetching statistics:", error);
  }
};

export const productsStatistics = async () => {
  try {
    const response = await api.get("/statistics/product");
    return response.data;
  } catch (error) {
    console.error("Error fetching products statistics:", error);
  }
};

export const employeesStatistics = async () => {
  try {
    const response = await api.get("/statistics/employees");
    return response.data;
  } catch (error) {
    console.error("Error fetching employees statistics:", error);
  }
};

export const weeklyStatistics = async () => {
  try {
    const response = await api.get("/statistics/weekly");
    return response.data;
  } catch (error) {
    console.error("Error fetching weekly statistics:", error);
  }
};
