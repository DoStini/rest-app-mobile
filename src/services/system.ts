import api from "./axios";
import * as Sentry from "sentry-expo";

export const fetchTablesInfo = async () => {
  try {
    const response = await api.get("/system/tables");
    return response.data;
  } catch (error) {
    console.error("Error fetching history:", error);
    Sentry.Native.captureException(error);
  }
};
