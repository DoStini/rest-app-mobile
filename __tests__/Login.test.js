import React from "react";
import { render, fireEvent, waitFor, act } from "@testing-library/react-native";
import Login from "../src/components/auth/Login";
import { login } from "../src/services/auth";

jest.mock("../src/services/auth");

describe("Login component", () => {
  const mockRevalidate = jest.fn();

  it("renders correctly", () => {
    const { getByTestId } = render(<Login revalidate={mockRevalidate} />);
    expect(getByTestId("email")).toBeTruthy();
    expect(getByTestId("password")).toBeTruthy();
    expect(getByTestId("submit")).toBeTruthy();
  });

  it("Snapshot is correct", () => {
    const { toJSON } = render(<Login revalidate={mockRevalidate} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("validates fields correctly", async () => {
    const { getByTestId, findByText } = render(
      <Login revalidate={mockRevalidate} />
    );

    fireEvent.changeText(getByTestId("email"), "invalid");
    fireEvent.changeText(getByTestId("password"), "");
    fireEvent.press(getByTestId("submit"));

    expect(await findByText("Invalid email")).toBeTruthy();
    expect(await findByText("Password is required")).toBeTruthy();
  });

  it("handles user input correctly", async () => {
    const { getByTestId } = render(<Login revalidate={mockRevalidate} />);
    fireEvent.changeText(getByTestId("email"), "test@example.com");
    fireEvent.changeText(getByTestId("password"), "password123");

    await waitFor(() => {
      expect(getByTestId("email").props.value).toBe("test@example.com");
      expect(getByTestId("password").props.value).toBe("password123");
    });
  });

  it("submits the form correctly", async () => {
    login.mockResolvedValueOnce();
    const { getByTestId } = render(<Login revalidate={mockRevalidate} />);

    fireEvent.changeText(getByTestId("email"), "user@example.com");
    fireEvent.changeText(getByTestId("password"), "password123");
    fireEvent.press(getByTestId("submit"));

    await waitFor(() => {
      expect(login).toHaveBeenCalledWith("user@example.com", "password123");
      expect(mockRevalidate).toHaveBeenCalled();
    });
  });

  it("handles loading and error states", async () => {
    login.mockRejectedValueOnce(new Error("Invalid credentials"));
    const { getByTestId, findByText } = render(
      <Login revalidate={mockRevalidate} />
    );

    fireEvent.changeText(getByTestId("email"), "user@example.com");
    fireEvent.changeText(getByTestId("password"), "wrongpassword");
    fireEvent.press(getByTestId("submit"));

    expect(await findByText("Invalid email or password")).toBeTruthy();
  });
});
