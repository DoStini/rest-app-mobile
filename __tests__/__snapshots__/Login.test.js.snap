// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component Snapshot is correct 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "maxWidth": 400,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 32,
        "fontWeight": "bold",
      }
    }
  >
    Welcome to Resty
  </Text>
  <Text
    style={
      {
        "fontSize": 14,
        "paddingBottom": 30,
        "paddingTop": 15,
      }
    }
  >
    Please authenticate to access the platform
  </Text>
  <View
    style={
      {
        "maxWidth": 400,
        "width": "80%",
      }
    }
  >
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Email"
      style={
        {
          "borderBottomColor": "#372F22",
          "borderBottomWidth": 1,
          "height": 40,
          "marginVertical": 12,
          "paddingVertical": 10,
        }
      }
      testID="email"
      value=""
    />
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Password"
      secureTextEntry={true}
      style={
        {
          "borderBottomColor": "#372F22",
          "borderBottomWidth": 1,
          "height": 40,
          "marginVertical": 12,
          "paddingVertical": 10,
        }
      }
      testID="password"
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      animatedStyle={
        {
          "value": {},
        }
      }
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#96744E",
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
      testID="submit"
    >
      <View
        style={
          {
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 16,
            }
          }
        >
          Login
        </Text>
      </View>
    </View>
  </View>
</View>
`;
